### PR TITLE
binary_search_tree.py: Add BST data structure

### PR DESCRIPTION
**Pull request by gabrieldluca**
This adds Binary Search Tree algorithm which keeps values in sorted
order, so that lookup and other operations can use the principles of
binary search while traversing the tree from root to leaf.

Closes https://github.com/NITSkmOS/Algorithms/issues/51

**By submitting this pull request I confirm I've read and complied with the
below declarations.**

- [X] I have read the [Contribution guidelines](https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [X] I have followed the [commit guidelines](https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md#write-good-commit-messages) for this project.
- [X] My code follows the [standard code structure](https://github.com/NITSkmOS/Algorithms/blob/master/CONTRIBUTING.md#samples-code).
- [X] This pull request has a descriptive title. For example, `{Tag}: Add {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [X] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.
- [X] This pull request shall only be reviewed and merged once all the checks passes. No maintainer or supporter shall be obliged to review it before this condition is met.
- [X] I have mentioned the issue number correctly (with hyperlink) in this pull request description.

After you submit your pull request, **DO NOT** click the 'Update Branch' button.

### CODE REVIEW

**Comment by gabrieldluca** [comment]
@sangamcse Please review.

**Comment by sangamcse** [comment]
@sangamcse rebase

**Comment by sangamcse** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by sangamcse** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by sangamcse** [comment]
Update in Readme and code comments needed.

File: binary_search_tree/Python/binary_search_tree.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+class Leaf:
+    """
+    Class that represent a leaf node of a binary tree.
+    """
+
```
Comment (sangamcse) [inline binary_search_tree/Python/binary_search_tree.py]: Line contains following spacing inconsistencies:
- Trailing whitespaces.

*Origin: SpaceConsistencyBear, Section: `python spacing`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
+++ b/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
@@ -2,7 +2,7 @@
     """
     Class that represent a leaf node of a binary tree.
     """
-    
+
     def __init__(self, value=None, leftChild=None, rightChild=None):
         self.value = value
         self.leftChild = leftChild
```

File: binary_search_tree/Python/binary_search_tree.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    def getChildren(self):
+        return [self.leftChild, self.rightChild]
+
+class BST:
+    """
+    Class that represents a Binary Search Tree
+    <https://en.wikipedia.org/wiki/Binary_search_tree>.
+    """
+
```
Comment (sangamcse) [inline binary_search_tree/Python/binary_search_tree.py]: Line contains following spacing inconsistencies:
- Trailing whitespaces.

*Origin: SpaceConsistencyBear, Section: `python spacing`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
+++ b/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
@@ -27,7 +27,7 @@
     Class that represents a Binary Search Tree
     <https://en.wikipedia.org/wiki/Binary_search_tree>.
     """
-    
+
     def __init__(self):
         self.root = None
         
```

File: binary_search_tree/Python/binary_search_tree.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+class BST:
+    """
+    Class that represents a Binary Search Tree
+    <https://en.wikipedia.org/wiki/Binary_search_tree>.
+    """
+    
+    def __init__(self):
+        self.root = None
+
```
Comment (sangamcse) [inline binary_search_tree/Python/binary_search_tree.py]: Line contains following spacing inconsistencies:
- Trailing whitespaces.

*Origin: SpaceConsistencyBear, Section: `python spacing`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
+++ b/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
@@ -30,7 +30,7 @@
     
     def __init__(self):
         self.root = None
-        
+
     def __del__(self):
         del self.root
 
```

File: binary_search_tree/Python/binary_search_tree.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        if currentNode is None:
+            return Leaf(value)
+        elif value < currentNode.value:
+            currentNode.leftChild = self.__insertNode(currentNode.leftChild, value)
+        elif value > currentNode.value:
+            currentNode.rightChild = self.__insertNode(currentNode.rightChild, value)
+        else:
+            currentNode.value = value
+        return currentNode
+
```
Comment (sangamcse) [inline binary_search_tree/Python/binary_search_tree.py]: Line contains following spacing inconsistencies:
- Trailing whitespaces.

*Origin: SpaceConsistencyBear, Section: `python spacing`.*

The issue can be fixed by applying the following patch:

```diff
--- a/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
+++ b/tmp/tmpikx6la4g/binary_search_tree/Python/binary_search_tree.py
@@ -60,7 +60,7 @@
         else:
             currentNode.value = value
         return currentNode
-        
+
 def main():
     values = [1, 2, 0, 1.5, 0.5]
     tree = BST()
```

**Comment by gabrieldluca** [comment]
@sangamcse I've updated the code problems and updated the README file. Please check again once you're ready.

**Comment by sangamcse** [comment]
@sangamcse rebase

**Comment by sangamcse** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by sangamcse** [comment]
Automated rebase failed! Please rebase your pull request manually via the command line.

Reason:
```
Command: git rebase base/master

Exit Code: 128

Cause:

.git/rebase-apply/patch:27: trailing whitespace.
class Leaf:
.git/rebase-apply/patch:28: trailing whitespace.
    """
.git/rebase-apply/patch:29: trailing whitespace.
    Class that represent a leaf node of a binary tree.
.git/rebase-apply/patch:30: trailing whitespace.
    """
.git/rebase-apply/patch:31: trailing whitespace.

warning: squelched 74 whitespace errors
warning: 79 lines add whitespace errors.
error: Failed to merge in the changes.


```

**Comment by sangamcse** [comment]
@gabrieldluca, I tried rebasing your repo but it failed. Kindly rebase it locally and push it again on success.

**Comment by gabrieldluca** [comment]
@sangamcse Done!

**Comment by sangamcse** [comment]
@sangamcse rebase

**Comment by sangamcse** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by sangamcse** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by sangamcse** [comment]
@sangamcse rebase

**Comment by sangamcse** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by sangamcse** [comment]
Automated rebase failed! Please rebase your pull request manually via the command line.

Reason:
```
Command: git rebase base/master

Exit Code: 128

Cause:

.git/rebase-apply/patch:20: trailing whitespace.
class Leaf:
.git/rebase-apply/patch:21: trailing whitespace.
    """
.git/rebase-apply/patch:22: trailing whitespace.
    Class that represent a leaf node of a binary tree.
.git/rebase-apply/patch:23: trailing whitespace.
    """
.git/rebase-apply/patch:24: trailing whitespace.

warning: squelched 74 whitespace errors
warning: 79 lines add whitespace errors.
error: Failed to merge in the changes.


```

**Pull request closed without merging**